### ISSUE TITLE
A contactpoint can get an invalid normal (0, 0, 0), which will cause an assert.

### ISSUE THREAD

**Issue by haikarainen** [DanielChappuis/reactphysics3d#32]
This happens when I create identical spheres in identical locations. 

See `ContactManifoldSet::computeCubemapNormalId` for example where it may assert:

```c++
short int ContactManifoldSet::computeCubemapNormalId(const Vector3& normal) const {

    assert(normal.lengthSquare() > MACHINE_EPSILON);

    // ...
```

I resolved this by patching `SphereVsSphereAlgorithm::testCollision`, adding a `Vector3 collisionNormal` variable that I made sure was sane, and used that instead of just the `vectorBetweenCenters.getUnit()`. It falls back to Y+ in case it's invalid.

There may be other collision algorithms that need fixes as well, I have only tested spheres.

See patched function here (too lazy to make a pull request, as my local copy has some substantial application-related changes):

```c++
void SphereVsSphereAlgorithm::testCollision(const CollisionShapeInfo& shape1Info,
                                            const CollisionShapeInfo& shape2Info,
                                            NarrowPhaseCallback* narrowPhaseCallback) {
    
    // Get the sphere collision shapes
    const SphereShape* sphereShape1 = static_cast<const SphereShape*>(shape1Info.collisionShape);
    const SphereShape* sphereShape2 = static_cast<const SphereShape*>(shape2Info.collisionShape);

    // Get the local-space to world-space transforms
    const Transform& transform1 = shape1Info.shapeToWorldTransform;
    const Transform& transform2 = shape2Info.shapeToWorldTransform;

    // Compute the distance between the centers
    Vector3 vectorBetweenCenters = transform2.getPosition() - transform1.getPosition();
    decimal squaredDistanceBetweenCenters = vectorBetweenCenters.lengthSquare();

    // Compute the sum of the radius
    decimal sumRadius = sphereShape1->getRadius() + sphereShape2->getRadius();
    
    // If the sphere collision shapes intersect
    if (squaredDistanceBetweenCenters <= sumRadius * sumRadius) {
        Vector3 centerSphere2InBody1LocalSpace = transform1.getInverse() * transform2.getPosition();
        Vector3 centerSphere1InBody2LocalSpace = transform2.getInverse() * transform1.getPosition();
        Vector3 intersectionOnBody1 = sphereShape1->getRadius() *
                                      centerSphere2InBody1LocalSpace.getUnit();
        Vector3 intersectionOnBody2 = sphereShape2->getRadius() *
                                      centerSphere1InBody2LocalSpace.getUnit();
        decimal penetrationDepth = sumRadius - std::sqrt(squaredDistanceBetweenCenters);
        
        Vector3 collisionNormal = vectorBetweenCenters.getUnit();
        if (collisionNormal.length() < MACHINE_EPSILON)
        {
          collisionNormal = Vector3(decimal(0.0), decimal(1.0), decimal(0.0)).getUnit();
        }

        // Create the contact info object
        ContactPointInfo contactInfo(shape1Info.proxyShape, shape2Info.proxyShape, shape1Info.collisionShape,
                                     shape2Info.collisionShape, collisionNormal, penetrationDepth,
                                     intersectionOnBody1, intersectionOnBody2);

        // Notify about the new contact
        narrowPhaseCallback->notifyContact(shape1Info.overlappingPair, contactInfo);
    }
}
```

**Comment by haikarainen** [comment]
Actually, I just stumbled into another scenario where zero-normal contact points causes issues. This time in `ContactSolver::computeFrictionVectors`.

Am I doing something wrong? Or is the library still very untested? Maybe I'm using an old version?

**Comment by haikarainen** [comment]
Anyway, I "solved" the second issue by patching `ContactSolver::initializeContactConstraints()`, by adding  the following code right before we call `computeFrictionVectors`:

```c++
            // Make sure normal is valid
            if (manifold.normal.length() < MACHINE_EPSILON)
            {
              manifold.normal = Vector3(0.0, 1.0, 0.0).getUnit();
            }
```

I'd love some feedback on this though, because it feels kinda hacky. Should this library be used?

**Comment by haikarainen** [comment]
This issue is starting to get very cluttered so I'll keep it short; Is it safe to assume that these asserts are something you use only for debug builds? Just realized they shouldn't crash the application in release builds..

**Comment by DanielChappuis** [comment]
Thanks a lot for reporting this. I have refactored a lot the collision detection system in the library for the next release (version 0.7.0) of the library (see branch "sat"). I need to check if those issues are still present with my refactoring and need to make sure that those bugs are fixed in the next release.

Concerning the asserts, yes I only use them in DEBUG builds to make sure that everything happens as expected.

**Comment by xSeditx** [comment]
I am having an issue which I believe may be similar to this one. I did not want to open my own Issue on here because Its more then likely my inexperience but its seeming that as soon as objects collide I get an assert and am not sure why... 

[code main.cpp] //attempting to Make the ground
// Initial position and orientation of the rigid body
rp3d :: Vector3 initPosition (-0.0 , -0.0 ,-0.0) ;
rp3d :: Quaternion initOrientation = rp3d :: Quaternion :: identity ();
rp3d :: Transform transform ( initPosition , initOrientation );
// Create a rigid body in the world
rp3d :: RigidBody *Ground ;
Ground = world.createRigidBody(transform);
BoxShape *box = new BoxShape(Vector3(500.f,20.f, 500.f));
Ground->addCollisionShape(box,transform,100.0);
Ground->setType(STATIC);
Ground->enableGravity ( false );

[code Sphere.cpp] // Generates a Vector of Sphere objects in the scene. 
Vector3      initPosition(Position.x, Position.y, Position.z) ;
Quaternion   initOrientation = Quaternion :: identity ();
Transform    transform ( initPosition , initOrientation );
// Create a rigid body in the world
Body = world.createRigidBody(transform);
Body->setType(DYNAMIC);
Shape = new SphereShape(Radius*2.5);
Body->addCollisionShape(Shape, transform,  10);

I believe this is the gist of my code, I am more then likely doing something wrong, but I have no other place to turn to ask for help. Any suggestions on why the Assert is happening
Assert : R6010
ContactPoint.cpp line 51
Its still early in me learning to use this so I am not even sure the Spheres I show on the screen are matching up with the collision objects just yet making debugging very hard

**Comment by DanielChappuis** [comment]
Hello,

Thanks for reporting this. I will release version 0.7.0 of the library very soon and I have refactored the collision detection a lot. This issue should be solved in the new version.

**Comment by DanielChappuis** [comment]
This issue should now be fixed in the version v0.7.0 of ReactPhysics3D that has just been released.
Thanks a lot for reporting this.
